collectors/carbonblack/utils.js

```diff
@@ -0,0 +1,171 @@
+    let nextPage;
+    if (apiDetails.method === "GET") {
+        return new Promise(function (resolve, reject) {
+            request({
+                url: apiDetails.url,
+                method: apiDetails.method,
+                headers: {
+                    'X-Auth-Token': `${clientSecret}/${clientId}`
+                },
+            }, (err, response) => {
+                if (err) {
+                    return reject(err);
+                }
+                response.body = JSON.parse(response.body);
+                if (response.body.notifications === undefined || response.body.notifications === null) {
+                    return reject(response.body.message);
+                }
+                if (response.body.notifications.length === 0) {
+                    return resolve({ accumulator, nextPage });
+                }
+                accumulator.push(...response.body.notifications);
+                return resolve({ accumulator, nextPage });
+            });
+
+        });
+    }
+    else {
+        let pageCount = 0;
+        let limit = 1;
+        let offset = state.nextPage ? state.nextPage : 1;
+        apiDetails.requestBody.rows = limit;
+        apiDetails.requestBody.start = offset;
+        return new Promise(function (resolve, reject) {
+            getCarbonBlackData();
+            function getCarbonBlackData() {
+                if (pageCount < maxPagesPerInvocation) {
+                    request({
+                        url: apiDetails.url,
+                        method: apiDetails.method,
+                        headers: {
+                            'X-Auth-Token': `${clientSecret}/${clientId}`
+                        },
+                        json: apiDetails.requestBody
+                    }, (err, response) => {
+                        if (err) {
+                            return reject(err);
+                        }
+                        if (response.body.results === undefined || response.body.results === null) {
+                            return reject(response.body.message);
+                        }
+                        if (response.body.results.length === 0) {
+                            return resolve({ accumulator, nextPage });
+                        }
+                        accumulator.push(...response.body.results);
+                        offset = offset + limit;
+                        apiDetails.requestBody.start = offset;
+                        pageCount++;
+                        return getCarbonBlackData();
+                    });
+                }
+                else {
+                    nextPage = offset;
+                    return resolve({ accumulator, nextPage });
+                }
+            }
+        });
+    }
+
+}
+
+function getAPIDetails(state, apiEndpoint, orgKey) {
+    let url = "";
+    let method = "GET";
+    let requestBody = "";
+    let typeIdPaths = [];
+    let tsPaths = [];
```

Comment:
See previous comment about pulling timestamp from eventTime field in the log message